erasurecoding/src/lib.rs

```diff
@@ -0,0 +1,600 @@
+use libc::{c_int, size_t};
+use std::slice;
+
+const MAX_SHARDS: usize = 65535;
+
+/// Reed-Solomon encode function for FFI
+/// Takes sharded original data as byte array and produces recovery shards
+/// Parameters:
+/// - original_shards: number of original data shards
+/// - original_shards_len: length of original_shards
+/// - recovery_shards: number of recovery shards to generate
+/// - original_shards: input byte array containing flattened shards
+/// - shard_size: size of each shard in bytes
+/// - recovery_shards_out: buffer to store generated and flattened recovery shards
+/// - recovery_shards_out_len: length of recovery_shards_out
+/// Returns 0 on success, -1 on error
+#[no_mangle]
+pub unsafe extern "C" fn reed_solomon_encode(
+    original_shards_count: size_t,
+    recovery_shards_count: size_t,
+    shard_size: size_t,
+    original_shards: *const u8,
+    original_shards_len: size_t,
+    recovery_shards_out: *mut u8,
+    recovery_shards_out_len: size_t,
+) -> c_int {
+    if let Some(sum) = original_shards_count.checked_add(recovery_shards_count) {
+        if sum > MAX_SHARDS {
+            return -1;
+        }
+    } else {
+        // Overflow.
+        return -1;
+    }
+
+    if !(original_shards_count > 0 && recovery_shards_count > 0) {
+        return -1;
+    }
+
+    if original_shards.is_null() || recovery_shards_out.is_null() {
+        return -1;
+    }
+
+    // Shard size must be a multiple of 2.
+    if !(shard_size > 0 && shard_size % 2 == 0) {
+        return -1;
+    }
+
+    if original_shards_len % shard_size != 0
+        || original_shards_len / shard_size != original_shards_count
+    {
+        return -1;
+    }
+
+    if recovery_shards_out_len != recovery_shards_count * shard_size {
+        return -1;
+    }
+
+    let original_shards =
+        slice::from_raw_parts(original_shards, original_shards_count * shard_size)
+            .chunks(shard_size);
+
+    match reed_solomon_simd::encode(
+        original_shards_count,
+        recovery_shards_count,
+        original_shards,
+    ) {
+        Ok(recovery) => {
+            let output_slice =
+                slice::from_raw_parts_mut(recovery_shards_out, recovery_shards_count * shard_size);
+
+            for (i, shard) in recovery.iter().enumerate() {
+                let start = i * shard_size;
+                let end = start + shard_size;
+                output_slice[start..end].copy_from_slice(shard);
```

Comment:
nit: `end` is basically not needed here. But if you like, keep it 😄 

```suggestion
                let start = i * shard_size;
                output_slice[start..start + shard_size].copy_from_slice(shard);
```